### PR TITLE
Sync our ctags to the latest tag version (p6.1.20240421.0)

### PR DESCRIPTION
**Pull request by techee**
There were mostly expected "problems" like new source files that had to be added, rename of `perl6.c` to `raku.c`, update of kind mappings, etc.

The only unexpected thing is the behavior of the javascript parser - this is from the javascript commit message:

There are lots of differences because of
    
https://github.com/universal-ctags/ctags/commit/6d85089456ed215ce6b6a673744ae42ccc5e0e99
    
Also
    
https://github.com/universal-ctags/ctags/commit/b1870b826a384c35671937743720464947af3b7e
    
seems to confuse the parser in simple.js so it doesn't generate `my_global_var2`.
    
Finally, Geany reports
```    
(geany:820768): Tagmanager-WARNING **: 20:38:28.755: ignoring null tag in /home/parallels/projects/geany/doc/reference/jquery.js(line: 2, language: JavaScript)
```

@b4n Does the PR in general and the javascript parser in particular look good to you? I didn't investigate the javascript parser differences much as I'm not a javascript user myself. At least the NULL warning should be fixed though.

### CODE REVIEW

**Comment by techee** [comment]
I also added a commit updating the matlab parser to the upstream regex-based one. I'd like to avoid the situation that we have to maintain our (non-regex) version of the parser like in:

https://github.com/geany/geany/pull/3358
https://github.com/geany/geany/pull/3563

Also, the kinds generated by the upstream parser are different than those we parse using our parser so the ctags file are incompatible.

**Comment by b4n** [comment]
I'll look into this Soon™, but the JS changes might have an issue (might be what you see, might be something else), see [this](https://github.com/universal-ctags/ctags/commit/6d85089456ed215ce6b6a673744ae42ccc5e0e99#diff-eb186936271438240da764e7ccf36ffa04ba8f0a6e1eabd1c69400164ae9f608R1) which doesn't look right to me.

In any case, thanks for doing this :)

**Comment by b4n** [comment]
Regarding JS:

> The only unexpected thing is the behavior of the javascript parser - this is from the javascript commit message:
> 
> There are lots of differences because of
> 
> [universal-ctags/ctags@6d85089](https://github.com/universal-ctags/ctags/commit/6d85089456ed215ce6b6a673744ae42ccc5e0e99)

Yeah, all this look OK.  Some can be viewed as implementor's choice, but it's fair so I'm fine with it.

> Also
> 
> [universal-ctags/ctags@b1870b8](https://github.com/universal-ctags/ctags/commit/b1870b826a384c35671937743720464947af3b7e)
> 
> seems to confuse the parser in simple.js so it doesn't generate `my_global_var2`.

Despite what's in the test file, this is *not* valid JS (AFAICT, or as far as SpiderMonkey tells me), so it's understandable the parser doesn't get along very well with it.  If you add a semicolon after it it fixes it, as it then consider it the end of the construct and "resets" itself.

> Finally, Geany reports
> 
> ```
> (geany:820768): Tagmanager-WARNING **: 20:38:28.755: ignoring null tag in /home/parallels/projects/geany/doc/reference/jquery.js(line: 2, language: JavaScript)
> ```

Is this actually anything new?  I just fixed a few upstream in https://github.com/universal-ctags/ctags/pull/3993, but I believe we already had this before.

Also, there's one instance in that jQuery code where it's somewhat expected, as there is indeed a zero-length property. For that remaining one, I'm not sure what the parser should do, as it can't really emit an empty name (format would probably not accept it, and anyway consumers are unlikely to like it), yet there *is* one in the input…

> @b4n Does the PR in general and the javascript parser in particular look good to you? I didn't investigate the javascript parser differences much as I'm not a javascript user myself. At least the NULL warning should be fixed though.

Yeah, the JS changes look (almost) fine.  I raised a couple issues upstream, but it's nothing serious nor that actually affect us much (see e.g. https://github.com/universal-ctags/ctags/issues/3995 and https://github.com/universal-ctags/ctags/issues/3997).

Two things:
* we should probabmy use a newer snapshot (some parser changes, e.g. related to class vs. object vs. variable are gonna lead to additional test result changes, which are admittedly better in the newer version, and I'm not sure it's worth having an intermediary step here)
* we should probably rather fix *simple.js* test to remove the invalid syntax that confuses the parser.  I'll try and do that Soon™.

**Comment by b4n** [comment]
BTW:

> Sync our ctags to the latest tag version (p6.1.20240421.0)

Tags `p*` are NOT releases, but probably pre-versions, or even nightly snapshots.  Releases are `v*`.
It's probably fine, but noteworthy :slightly_smiling_face:

**Comment by techee** [comment]
@b4n Thanks for having a look at the javascript parser.

> we should probabmy use a newer snapshot (some parser changes, e.g. related to class vs. object vs. variable are gonna lead to additional test result changes, which are admittedly better in the newer version, and I'm not sure it's worth having an intermediary step here)

Can do that. Should it be part of this PR or a separate one after this one is merged?

> we should probably rather fix simple.js test to remove the invalid syntax that confuses the parser. I'll try and do that Soon™.

Could do it myself but better someone who at least remotely knows what the correct syntax is :-).

> Tags p* are NOT releases, but probably pre-versions, or even nightly snapshots. Releases are v*.
It's probably fine, but noteworthy 🙂

OK, I didn't notice the release tags. But since these are just about once a year, I guess we want the latest thing and "risk" the consequences, right?

File: src/tagmanager/tm_parser.c [hunk 1]
```diff
@@ -1135,6 +1154,17 @@ static TMParserMapGroup group_OCAML[] = {
...
+/* unused by Geany but used by the cxx parser as an external parser so it has to
+ * be defined */
+static TMParserMapEntry map_LDSCRIPT[] = {
+	{'S', tm_tag_undef_t},  // section
+	{'s', tm_tag_undef_t},  // symbol
+	{'v', tm_tag_undef_t},  // version
+	{'i', tm_tag_undef_t},  // inputSection
+};
+static TMParserMapGroup group_LDSCRIPT[] = {
+};
```
Comment (b4n) [inline src/tagmanager/tm_parser.c:1166]: This gives me:
```
src/tagmanager/tm_parser.c:1165:44: warning: ISO C forbids empty initializer braces [-Wpedantic]
 1165 | static TMParserMapGroup group_LDSCRIPT[] = {
      |                                            ^
src/tagmanager/tm_parser.c:1165:25: error: zero or negative size array 'group_LDSCRIPT'
 1165 | static TMParserMapGroup group_LDSCRIPT[] = {
```
Might be related to my pedantic flags, but that's a bit problematic.

What about having actual groups, even if unused?
```suggestion
static TMParserMapGroup group_LDSCRIPT[] = {
	{N_("Section"), TM_ICON_STRUCT, tm_tag_struct_t},
	{N_("Symbol"), TM_ICON_METHOD, tm_tag_method_t | tm_tag_function_t},
	{N_("Version"), TM_ICON_OTHER, tm_tag_other_t},
	{N_("Input Section"), TM_ICON_NAMESPACE, tm_tag_namespace_t},
};
```

Or just a dummy one?
```suggestion
static TMParserMapGroup group_LDSCRIPT[] = {
	{"unused", TM_ICON_NONE, tm_tag_undef_t},
};
```

File: tests/ctags/matlab_test.m.tags
Comment (b4n) [inline tests/ctags/matlab_test.m.tags:1]: that's a bit sad… I'll see if I can get those fixed upstream (it's easy enough to patch actually -- yet, I don't have a clue about matlab :grin: )

File: src/tagmanager/tm_parser.c [hunk 2]
```diff
@@ -557,6 +562,7 @@ static TMParserMapEntry map_FREEBASIC[] = {
 	{'t', tm_tag_struct_t},     // type
 	{'v', tm_tag_variable_t},   // variable
 	{'g', tm_tag_externvar_t},  // enum
+	{'n', tm_tag_undef_t},      // namespace
```
Comment (b4n) [inline src/tagmanager/tm_parser.c:565]: No clue about freebasic, but wouldn't it be worth including?  kinds are already silly, so the fact `tm_tag_namespace_t` is already in use is not really a reason not to add it, is it?

**Comment by b4n** [comment]
> Can do that. Should it be part of this PR or a separate one after this one is merged?

Whichever you prefer.  I'd have said this one given it'll introduce yet another set of changes that are probably best diffed from current master, but I don't really mind either way in practice, so if it's easier for you just do another one afterwards.

> > Tags p* are NOT releases, but probably pre-versions, or even nightly snapshots. Releases are v*.
> > It's probably fine, but noteworthy 🙂
> 
> OK, I didn't notice the release tags. But since these are just about once a year, I guess we want the latest thing and "risk" the consequences, right?

Yeah, I think we can "risk it" :slightly_smiling_face:

File: src/tagmanager/tm_parser.c (hunk 1 above)
Comment (techee) [reply src/tagmanager/tm_parser.c]: I added the dummy one in the latest commit.

File: src/tagmanager/tm_parser.c (hunk 2 above)
Comment (techee) [reply src/tagmanager/tm_parser.c]: Yes, these might be missing in the hierarchy. Added them in the latest commit including the unit test I copied from universal-ctags.

File: src/tagmanager/tm_parser.c (hunk 1 above)
Comment (b4n) [reply src/tagmanager/tm_parser.c]: BTW, it's not merely a warning, but I had an error.  However, it's a [GCC extension](https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html)

**Comment by b4n** [comment]
@techee I just added 2 more commits -- feel free to comment and reject them :)
Anyway, looks good to me now.

**Review by b4n — APPROVED** [review]
I didn't review all parser's changes (sorry :sob:), but the Geany changes, tests and behavior look good :+1:

**Comment by techee** [comment]
> @techee I just added 2 more commits -- feel free to comment and reject them :)

Yes, that's what I'm gonna do, just to enjoy the power :-).

Anyway, great, I've merged this PR. If you plan to do some more ctags changes, just ping me when you are done and the changes are merged upstream and I'll prepare another sync PR with them.

**Comment by b4n** [comment]
@techee FWIW, the PRs I wanted merged now are, so it's whenever you like :)

**Comment by techee** [comment]
Great, I think I'll just wait for the next ctags tag and will use that one.

**Pull request merged**